File: src/main/java/com/thebluehats/server/game/managers/enchants/CustomEnchantUtils.java
package com.thebluehats.server.game.managers.enchants;

import java.util.List;

import com.google.inject.Inject;
import com.thebluehats.server.game.utils.RomanNumeralConverter;

import org.bukkit.ChatColor;
import org.bukkit.Material;
import org.bukkit.inventory.ItemStack;

public class CustomEnchantUtils {
    private final RomanNumeralConverter romanNumeralConverter;

    @Inject
    public CustomEnchantUtils(RomanNumeralConverter romanNumeralConverter) {
        this.romanNumeralConverter = romanNumeralConverter;
    }

    public boolean isCompatibleWith(CustomEnchant enchant, Material material) {
        for (Material mat : enchant.getEnchantItemTypes()) {
            if (mat == material) {
                return true;
            }
        }

        return false;
    }

    public boolean itemHasEnchant(CustomEnchant enchant, ItemStack item) {
        if (item == null || item.getType() == Material.AIR)
            return false;

        List<String> lore = item.getItemMeta().getLore();

        if (lore == null)
            return false;

        for (String line : lore) {
            if (line.contains(enchant.getName())) {
                return true;
            }
        }

        return false;
    }

    public int getEnchantLevel(CustomEnchant enchant, ItemStack item) {
        if (item == null || item.getType() == Material.AIR)
            return 0;

        List<String> lore = item.getItemMeta().getLore();

        if (lore == null)
            return 0;

        for (String line : lore) {
            if (line.contains(enchant.getName())) {
                String strippedLine = ChatColor.stripColor(line);

                String[] lineTokens = strippedLine.split(" ");
                String numeral = lineTokens[lineTokens.length - 1];

                return numeral == null ? 1 : romanNumeralConverter.convertRomanNumeralToInteger(numeral);
            }
        }

        return 0;
    }
}

File: src/main/kotlin/com/thebluehats/server/game/managers/enchants/CustomEnchantUtils.kt
package com.thebluehats.server.game.managers.enchants

import com.thebluehats.server.game.utils.numeralToInteger
import org.bukkit.ChatColor
import org.bukkit.Material
import org.bukkit.inventory.ItemStack

class CustomEnchantUtils {
    fun isCompatibleWith(enchant: CustomEnchant, material: Material): Boolean {
        return enchant.enchantItemTypes.any { t -> t == material }
    }

    fun getItemEnchantData(enchant: CustomEnchant, item: ItemStack?): ItemEnchantData {
        return ItemEnchantData(itemHasEnchant(enchant, item), getEnchantLevel(enchant, item))
    }

    fun itemHasEnchant(enchant: CustomEnchant, item: ItemStack?): Boolean {
        if (item == null || item.type == Material.AIR) return false
        val lore = item.itemMeta.lore ?: return false

        return lore.any { l -> l.contains(enchant.name) }
    }

    fun getEnchantLevel(enchant: CustomEnchant, item: ItemStack?): Int {
        if (item == null || item.type == Material.AIR) return 0
        val lore = item.itemMeta.lore ?: return 0

        lore.forEach { l ->
            if (!l.contains(enchant.name)) return@forEach

            val tokens = ChatColor.stripColor(l).split(" ").toTypedArray()
            val numeral = tokens[tokens.size - 1]

            val level = numeral.numeralToInteger()

            return if (level == -1) 1 else level
        }

        return 0
    }

    class ItemEnchantData constructor(private val itemHasEnchant: Boolean, val enchantLevel: Int) {
        fun itemHasEnchant(): Boolean {
            return itemHasEnchant
        }
    }
}
